db/chained_delegating_reader.hh

```diff
@@ -28,15 +28,17 @@
 // A reader which allows to insert a deferring operation before reading.
 // All calls will first wait for a future to resolve, then forward to a given underlying reader.
 class chained_delegating_reader : public flat_mutation_reader::impl {
-    flat_mutation_reader _underlying;
+    std::unique_ptr<flat_mutation_reader> _underlying;
     shared_future<> _ready;
     std::function<void()> _on_destroyed;
     
 public:
-    chained_delegating_reader(future<> ready, reader_permit permit, flat_mutation_reader underlying, std::function<void()> on_destroyed = []{})
-        : impl(underlying.schema(), std::move(permit))
-        , _underlying(std::move(underlying))
-        , _ready(std::move(ready))
+    chained_delegating_reader(schema_ptr s, std::function<future<flat_mutation_reader>()> populate, reader_permit permit, std::function<void()> on_destroyed = []{})
+        : impl(s, std::move(permit))
+        , _ready(populate().then([this] (flat_mutation_reader &&rd) {
```

Comment:
If you capture "this", you should delete the move constructor so that the object is not moved somewhere else.